spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/dataAdapter/GeoPackageAdapter.java

```diff
@@ -47,114 +81,552 @@
+
+                adapterInstance.setStatus(SCDataAdapterStatus.DATA_ADAPTER_CONNECTING);
+
+                URL theUrl = null;
+                try {
+                    // TODO: we need to determine if the store's URI is local to the filesystem (packaged
+                    // with the app or on an SD card and try loading from there)
+                    theUrl = new URL(scStoreConfig.getUri());
+                }
+                catch (MalformedURLException e) {
+                    Log.e(LOG_TAG, "URL was malformed. Check the syntax: " + theUrl);
+                    adapterInstance.setStatus(SCDataAdapterStatus.DATA_ADAPTER_DISCONNECTED);
+                    subscriber.onError(e);
+                }
+
+                // download geopackage and store it locally if it's not present
+                if (context.getDatabasePath(scStoreConfig.getName()).exists()) {
+                    Log.d(LOG_TAG, "GeoPackage " + scStoreConfig.getName() + " already exists.  Not downloading.");
+                    // create new GeoPackage for the file that's already on disk
+                    gpkg = new GeoPackage(context, scStoreConfig.getName());
+                    subscriber.onCompleted();
+                }
+                else {
+                    downloadGeoPackage(theUrl)
+                            .subscribeOn(Schedulers.io())
+                            .subscribe(new Action1<Response>() {
+                                @Override
+                                public void call(Response response) {
+                                    if (!response.isSuccessful()) {
+                                        subscriber.onError(new IOException("Unexpected code " + response));
+                                    }
+                                    try {
+                                        // save the file to the databases directory
+                                        saveFileToFilesystem(response.body().byteStream());
+                                        // create new GeoPackage now that we've saved the file
+                                        gpkg = new GeoPackage(context, scStoreConfig.getName());
+                                        adapterInstance.setStatus(SCDataAdapterStatus.DATA_ADAPTER_CONNECTED);
+                                        subscriber.onCompleted();
+                                    }
+                                    catch (IOException e) {
+                                        subscriber.onError(e);
+                                    }
+                                    finally {
+                                        response.body().close();
+                                    }
+                                }
+                            });
+                }
+            }
+        });
+    }
+
+
+    /**
+     * For each query, we use a <a href="https://en.wikipedia.org/wiki/Correlated_subquery">correlated sub query</a>
```

Comment:
Add note to the javadoc that the subscriber must unsubscribe to terminate the subscription, otherwise the query will not return onCompleted